File: python/dsa/algorithms/strings/longest-palindromic-substring.py
def longestPalindrome(s: str) -> str:
  if len(s) < 2: return s
  res = [0, 1] # start, maxLen
  
  for i in range(len(s)):
    expand(s, i, i, res)
    expand(s, i, i + 1, res)
    
  return s[res[0] : res[0] + res[1]]

def expand(s, l, r, res):
  while l >= 0 and r < len(s) and s[l] == s[r]:
    if (r - l + 1) > res[1]:
      res[1] = r - l + 1
      res[0] = l
    l -= 1
    r += 1

File: go/dsa/algorithms/strings/longest-palindromic-substring.go
func longestPalindrome(s string) string {
  if len(s) < 2 { return s }
  start, maxLen := 0, 1
  
  for i := 0; i < len(s); i++ {
    start, maxLen = expand(s, i, i, start, maxLen)
    start, maxLen = expand(s, i, i+1, start, maxLen)
  }
  return s[start : start+maxLen]
}

func expand(s string, l, r, start, maxLen int) (int, int) {
  for l >= 0 && r < len(s) && s[l] == s[r] {
    if r-l+1 > maxLen {
      maxLen = r-l+1
      start = l
    }
    l-- ; r++
  }
  return start, maxLen
}